File: backend/src/merlins_collection/services/acquisition.py
"""Acquisition economics — RFC 0024 T1.

One authority for the acquisition ratio, mirrored (not shared over the wire)
by ``frontend/lib/acquisition.ts``'s ``acquisitionRatio``. The two are pinned
together by a shared fixture — see ``backend/tests/test_cross_boundary.py``
and ``frontend/lib/__tests__/acquisition.test.ts``, both of which load
``shared/test-fixtures/acquisition-ratio-cases.json``.

``acquisition_ratio`` is deliberately NOT stored on ``InventoryItem``. It is
derived from two stored inputs (``market_value_at_purchase``, ``cost_basis``)
and would go stale the moment either changes — including from this same RFC's
own ``cost_basis`` sync on a transaction edit. Compute it at read time, every
time.
"""

from decimal import ROUND_HALF_UP, Decimal

_TWO_PLACES = Decimal("0.01")


def acquisition_ratio(
    market_value_at_purchase: Decimal | None,
    cost_basis: Decimal | None,
) -> Decimal | None:
    """``market_value_at_purchase / cost_basis``, as a PERCENT.

    The owner's "market at purchase / amount paid" — 312.50 means we paid $32
    for a card the market said was worth $100 at the time.

    ``None`` when either figure is absent, or when ``cost_basis`` is zero. A
    free card (a throw-in, a bulk lot) is a real and routine thing at a buy
    table, and its ratio is not "infinite" or "0" — it is undefined, and
    rendering either number would be a claim nobody made. Every caller must
    handle ``None`` and render an em dash.

    Rounded to two decimal places (``ROUND_HALF_UP``) so cross-language
    precision differences between Python's arbitrary-precision ``Decimal``
    and JavaScript's float64 ``Number`` cannot produce a spurious mismatch on
    a repeating-decimal division (e.g. 10/3) — both sides round to the same
    bounded precision before anything compares them.

    A negative ``cost_basis`` is out of scope: nothing upstream can produce
    one (``parseMoney`` and the backend's own money fields reject negative
    amounts before a value ever reaches storage), so this function only
    special-cases zero, not sign.
    """
    if market_value_at_purchase is None or cost_basis is None:
        return None
    if cost_basis == 0:
        return None

    raw = (market_value_at_purchase / cost_basis) * 100
    return raw.quantize(_TWO_PLACES, rounding=ROUND_HALF_UP)

File: frontend/lib/acquisition.ts
/**
 * Acquisition economics — RFC 0024 T1.
 *
 * The TypeScript mirror of
 * `backend/src/merlins_collection/services/acquisition.py`'s
 * `acquisition_ratio`. Two implementations, deliberately, in the same shape
 * as `itemTitle` / `adminItemName` / `admin_item_name` / MCP's `toCard` — one
 * rule, several implementations, kept in sync on purpose. The two are pinned
 * together by a shared fixture rather than by trust: see
 * `frontend/lib/__tests__/acquisition.test.ts` and
 * `backend/tests/test_cross_boundary.py`, both of which load
 * `shared/test-fixtures/acquisition-ratio-cases.json`.
 *
 * `acquisitionRatio` is never stored — it is derived from two figures already
 * on the wire (`market_value_at_purchase`, `cost_basis`) and would go stale
 * the moment either changes. Compute it at render time, every time.
 */

function toAmount(value: string | number | null | undefined): number | null {
  if (value === null || value === undefined || value === '') return null
  const n = typeof value === 'string' ? Number(value) : value
  return Number.isFinite(n) ? n : null
}

/**
 * `marketValueAtPurchase / costBasis`, as a PERCENT.
 *
 * `null` when either figure is absent, or when `costBasis` is zero — a free
 * card (a throw-in, a bulk lot) is routine at a buy table, and its ratio is
 * undefined, not infinite and not zero. Every caller must handle `null` and
 * render an em dash (or, for a chip, render nothing at all — see
 * `ratioTone`).
 *
 * Rounded to two decimal places, the same precision
 * `acquisition_ratio` rounds to on the Python side, so a repeating-decimal
 * division (e.g. 10/3) can't drift between float64 and arbitrary-precision
 * `Decimal` before the two are compared.
 */
export function acquisitionRatio(
  marketValueAtPurchase: string | number | null | undefined,
  costBasis: string | number | null | undefined
): number | null {
  const market = toAmount(marketValueAtPurchase)
  const cost = toAmount(costBasis)
  if (market === null || cost === null || cost === 0) return null

  const raw = (market / cost) * 100
  return Math.round(raw * 100) / 100
}

/**
 * The rounded integer percent for display, e.g. `"313%"`. `null` in, `null`
 * out — the caller renders nothing for a null ratio, never a grey `0%` or a
 * dash mixed into a chip; see `ratioTone`'s docstring for why a `null` chip
 * is not a chip at all.
 */
export function formatRatio(ratio: number | null): string | null {
  if (ratio === null) return null
  return `${Math.round(ratio)}%`
}

export type RatioTone = 'good' | 'neutral' | 'bad'

/**
 * Tone bands, defined once and nowhere else:
 *
 * | Ratio      | Tone                        |
 * |------------|-----------------------------|
 * | >= 200%    | good                        |
 * | 100 - 200% | neutral                     |
 * | < 100%     | bad — we paid over market   |
 * | `null`     | no chip at all, not a grey zero |
 */
export function ratioTone(ratio: number | null): RatioTone | null {
  if (ratio === null) return null
  if (ratio >= 200) return 'good'
  if (ratio >= 100) return 'neutral'
  return 'bad'
}
